File: LLM-Translations/GPT-4/codenet/C/Python/s427349585.py
def merge(a, n, l, m, r):
    L = a[l:m] + [{'s': '', 'v': 2000000000}]
    R = a[m:r] + [{'s': '', 'v': 2000000000}]
    i = j = 0
    for k in range(l, r):
        if L[i]['v'] <= R[j]['v']:
            a[k] = L[i]
            i += 1
        else:
            a[k] = R[j]
            j += 1

def merge_sort(a, n, l, r):
    if l + 1 < r:
        m = (l + r) // 2
        merge_sort(a, n, l, m)
        merge_sort(a, n, m, r)
        merge(a, n, l, m, r)

def partition(a, n, p, r):
    x = a[r]
    i = p - 1
    for j in range(p, r):
        if a[j]['v'] <= x['v']:
            i += 1
            a[i], a[j] = a[j], a[i]
    a[i + 1], a[r] = a[r], a[i + 1]
    return i + 1

def quick_sort(a, n, p, r):
    if p < r:
        q = partition(a, n, p, r)
        quick_sort(a, n, p, q - 1)
        quick_sort(a, n, q + 1, r)

def main():
    n = int(input())
    a = [0] * n
    b = [0] * n

    for i in range(n):
        s, v = input().split()
        v = int(v)
        a[i] = b[i] = {'s': s, 'v': v}

    merge_sort(a, n, 0, n)
    quick_sort(b, n, 0, n - 1)

    stable = 1
    for i in range(n):
        if a[i]['s'] != b[i]['s']:
            stable = 0

    if stable == 1:
        print("Stable")
    else:
        print("Not stable")

    for i in range(n):
        print(b[i]['s'], b[i]['v'])

main()

File: LLM-Translations/GPT-4/codenet/C/Java/s427349585.java
import java.util.Scanner;

class Card {
    char s;
    int v;

    Card(char s, int v) {
        this.s = s;
        this.v = v;
    }
}

public class s427349585 {
    static Card[] L = new Card[100000 / 2 + 2];
    static Card[] R = new Card[100000 / 2 + 2];

    static void merge(Card[] a, int n, int l, int m, int r) {
        int i, j, k;
        int n1 = m - l;
        int n2 = r - m;

        for (i = 0; i < n1; i++) L[i] = a[l + i];
        for (i = 0; i < n2; i++) R[i] = a[m + i];
        L[n1] = new Card('\0', 2000000000);
        R[n2] = new Card('\0', 2000000000);
        i = 0;
        j = 0;
        for (k = l; k < r; k++) {
            if (L[i].v <= R[j].v) {
                a[k] = L[i++];
            } else {
                a[k] = R[j++];
            }
        }
    }

    static void mergeSort(Card[] a, int n, int l, int r) {
        int m;
        if (l + 1 < r) {
            m = (l + r) / 2;
            mergeSort(a, n, l, m);
            mergeSort(a, n, m, r);
            merge(a, n, l, m, r);
        }
    }

    static int partition(Card[] a, int n, int p, int r) {
        int i, j;
        Card t, x;
        x = a[r];
        i = p - 1;
        for (j = p; j < r; j++) {
            if (a[j].v <= x.v) {
                i++;
                t = a[i];
                a[i] = a[j];
                a[j] = t;
            }
        }
        t = a[i + 1];
        a[i + 1] = a[r];
        a[r] = t;
        return i + 1;
    }

    static void quickSort(Card[] a, int n, int p, int r) {
        int q;

        if (p < r) {
            q = partition(a, n, p, r);
            quickSort(a, n, p, q - 1);
            quickSort(a, n, q + 1, r);
        }
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n, i, v;
        Card[] a = new Card[200000];
        Card[] b = new Card[200000];
        String s;
        boolean stable = true;

        n = sc.nextInt();

        for (i = 0; i < n; i++) {
            s = sc.next();
            v = sc.nextInt();
            a[i] = b[i] = new Card(s.charAt(0), v);
        }

        mergeSort(a, n, 0, n);
        quickSort(b, n, 0, n - 1);

        for (i = 0; i < n; i++) {
            if (a[i].s != b[i].s) {
                stable = false;
            }
        }
        if (stable) {
            System.out.println("Stable");
        } else {
            System.out.println("Not stable");
        }
        for (i = 0; i < n; i++) {
            System.out.printf("%c %d\n", b[i].s, b[i].v);
        }

        sc.close();
    }
}
// 
